Chore/simplify tools

vrknetha opened the pull request:
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [x] Refactoring (no functional changes)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

vrknetha committed:
chore: simplify tools.ts implementation and remove tool-specific logic

vrknetha committed:
Implement recursive schema validation for nested objects and arrays

vrknetha committed:
Add graceful fallback to console-only logging when file permissions unavailable

vrknetha committed:
Add @langchain/openai as a development dependency

vrknetha committed:
Fix test suite issues and add comprehensive test coverage for schema validation, nested schemas, performance tracking, and error handling

vrknetha committed:
fix: removes unwanted mcp json

vrknetha committed:
fix: removes unwanted mcp json

vbarda commented on README.md:
```diff
@@ -93,47 +93,31 @@ if __name__ == "__main__":
-  args: ["math_server.py"]
-});
+// Create client and connect to server using the adapter
+const client = new MultiServerMCPClient();
```
please revert this to the current README - we want to show how people can use w/ MCP SDK

vrknetha replied:
Reverted it

vbarda commented on examples/firecrawl_custom_config_example.ts:
```diff
@@ -0,0 +1,183 @@
+/**
```
can we change this to just use langgraph's `createReactAgent`? it does exactly the same thing under the hood

vrknetha replied:
Done

vbarda commented on examples/firecrawl_default_config_example.ts:
```diff
@@ -0,0 +1,166 @@
+/**
```
same comment as above re: `createReactAgent` - and it will also make it easier to highlight the differences between custom config / default config

vrknetha replied:
Changed it to createReactAgent

vbarda commented on examples/firecrawl_enhanced_config_example.ts:
```diff
@@ -0,0 +1,232 @@
+/**
```
same comment re: createReactAgent

vrknetha replied:
Changed

vbarda replied:
seems like it's still using a full `StateGraph` here - i think it's fine to keep at most one example using full state graph (and we already do in  
`examples/filesystem_langgraph_example.ts`) but would keep the rest using `createReactAgent`

vbarda commented on examples/firecrawl_mixed_loading_example.ts:
```diff
@@ -0,0 +1,197 @@
+/**
```
same comment

vrknetha replied:
Changed

vbarda commented on examples/firecrawl_multiple_servers_example.ts:
```diff
@@ -0,0 +1,188 @@
+/**
```
same comment

vrknetha replied:
Changed

vbarda commented on src/tools.ts:
```diff
@@ -27,420 +38,364 @@ class ToolException extends Error {
  */
-function _convertCallToolResult(result: CallToolResult): any {
+function _convertCallToolResult(result: CallToolResult): unknown {
   logger.debug('Processing MCP tool result:', JSON.stringify(result, null, 2));
```
is there a way to disable logging? some verbosity parameter people can set? i'd prefer that it was disabled by default and was an opt-in behavior

vrknetha replied:
I have disabled logging by default and updated readme about how to turn it on

vbarda commented on src/tools.ts:
```diff
@@ -27,420 +38,364 @@ class ToolException extends Error {
-  if (typeof input === 'object' && input !== null && Object.keys(input).length > 0) {
-    logger.debug('Input is already an object, returning as-is');
-    return input;
+function fixArrayProperties(obj: SchemaPropertyWithContext): void {
```
can we get rid of this?

vrknetha replied:
refactored it altogether

vbarda commented on src/tools.ts:
```diff
@@ -27,420 +38,364 @@ class ToolException extends Error {
+ * @param input - The input to normalize
+ * @returns Normalized input as a record
+ */
+function _normalizeInput(input: unknown): Record<string, unknown> {
```
can we get rid of this?

vrknetha replied:
This is refactored altogether

vrknetha committed:
git log -1refactor: simplify MCP tools implementation
git branch --show-current
git push# Please enter the commit message for your changes. Lines starting

vrknetha committed:
chore: removes unwanted dependency

vrknetha committed:
chore: changes console to logger

vrknetha committed:
fix: implement proper handling of content_and_artifact format in MCPTool class

vrknetha committed:
refactor: improvements based on PR feedback - Update README with MCP SDK client example - Set stdio as default transport for better DX - Make logging opt-in by default - Simplify example files

vrknetha committed:
refactor: convert firecrawl_multiple_servers_example to use createReactAgent

vrknetha committed:
docs: improve configuration management documentation in README

vbarda commented on README.md:
````diff
@@ -282,6 +284,8 @@ Define your server connections in a JSON file:
 }
 ```
 
+Note: For stdio connections, the `transport` field is optional. If not specified, it defaults to 'stdio'.
````
consider changing to this style to make the callout more prominent https://github.com/orgs/community/discussions/16925

vrknetha replied:
Done

vbarda commented on README.md:
````diff
@@ -296,6 +300,170 @@ await client.initializeConnections();
+}
+```
+
+> **Note:** For stdio connections, the `transport` field is optional. If not specified, it defaults to 'stdio' when `command` and `args` are present.
````
consider changing to this style to make the callout more prominent https://github.com/orgs/community/discussions/16925

vbarda commented on __tests__/tools.test.ts:
```diff
@@ -0,0 +1,255 @@
+import { Client } from '@modelcontextprotocol/sdk/client/index.js';
+import { convertMcpToolToLangchainTool, loadMcpTools, ResponseFormat } from '../src/tools.js';
```
looks like we're never using the import:

`'ResponseFormat' is defined but never used. Allowed unused vars must match /^_/u`

vrknetha replied:
Removed

vbarda commented on package.json:
```diff
@@ -38,7 +39,6 @@
   "author": "Ravi Kiran Vemula",
   "license": "MIT",
   "dependencies": {
-    "@dmitryrechkin/json-schema-to-zod": "^1.0.1",
```
nice!

vbarda commented on src/tools.ts:
```diff
@@ -27,718 +32,146 @@ class ToolException extends Error {
-      }
+  tool: { name: string; description?: string; inputSchema?: any },
+  toolSchema?: any,
+  responseFormat: ResponseFormat = 'text'
```
should the default actually be `content_and_artifact`? the only thing it will change is the `toolMessage.artifact` field after the tool is invoked - content will be unchanged

vrknetha replied:
done

vbarda commented on src/tools.ts:
```diff
@@ -27,718 +32,146 @@ class ToolException extends Error {
- * @param toolSchema - The schema of the tool
- * @returns A LangChain tool implementing StructuredToolInterface
+ * @param tool - The MCP tool to convert
+ * @param toolSchema - Tool schema (unused in simplified implementation)
```
do we need this? or is this kept for backwards compatibility? if so, can we indicate that somehow?

vrknetha replied:
Changed the comment

vbarda reviewed (approved):
looks great, thanks! added a few comments

vrknetha committed:
fix: address PR feedback on tools and tests
